Add lexer for Soong (Android.bp) config files

JonathonReinhart opened the pull request:
This adds a lexer for `Android.bp` files which are used by the *Soong* Android build system.

https://source.android.com/docs/setup/build

Here is an example of a full `Android.bp` file: [`platform/frameworks/base/Android.bp`](https://android.googlesource.com/platform/frameworks/base/+/55afce9107ea882fe236acc5653aef9d6fb2185b/Android.bp).

JonathonReinhart commented:
Hi @Anteru @jeanas, friendly ping! Is there anything I can do to advance this?

Anteru commented:
Just came back from travel, looking into this now.

JonathonReinhart commented:
Thanks @Anteru. I fixed up the `check` and `regexlint` errors.

Anteru commented on pygments/lexers/soong.py:
```diff
@@ -0,0 +1,70 @@
+        'common': [
+            (r'//.*?$', Comment.Single),
+            (r'/(\\\n)?[*](.|\n)*?[*](\\\n)?/', Comment.Multiline),
+            (r'\n', Whitespace),
```
`\n+`? Should be no problem merging consecutive whitespace characters.

JonathonReinhart replied:
Replaced with `\s+`.

Anteru commented on pygments/lexers/soong.py:
```diff
@@ -0,0 +1,70 @@
+            (r'//.*?$', Comment.Single),
+            (r'/(\\\n)?[*](.|\n)*?[*](\\\n)?/', Comment.Multiline),
+            (r'\n', Whitespace),
+            (r' +', Whitespace),
```
Does this language not support `\t` as a whitespace character?

JonathonReinhart replied:
Replaced with `\s+`.

Anteru commented on pygments/lexers/soong.py:
```diff
@@ -0,0 +1,70 @@
+            (r'0x[0-9a-fA-F]+', Number.Hex),
+            (r'\d+', Number.Integer),
+            (r'".*?"', String),
+            (r'[\[\]]', Punctuation),
+            (r',', Punctuation),
```
You can merge those two rules and use `+` to allow consuming multiple punctuation characters in one go.

JonathonReinhart replied:
Obsolete. Square brackets actually move to a new `in-list` state.

JonathonReinhart commented:
Thanks for the review, @Anteru! Your comment about whitespace helped me realize there was a deficiency in my previous version (multi-line lists weren't handled correctly). I've corrected this and added a test.

Please take another look!

jeanas commented on pygments/lexers/soong.py:
```diff
@@ -0,0 +1,77 @@
+            (r'\{', Punctuation, 'in-map'),
+            (r'\[', Punctuation, 'in-list'),
+            (r'\w+', Name),
+            include('comments'),
```
Please add a rule

```python
default('#pop'),
```

here. IIUC, it's working right now because we have an undocumented fallback to restarting in the `'root'` state when nothing matches and there's a newline boundary, but we might want to remove that, so better not rely on it...

JonathonReinhart replied:
Unfortunately, I don't think it's that simple. I tried that and it broke everything.

Currently, `expr` is used with both an `include()` and a new state. (That seems like a bad idea, now that I think about it.)

So while it makes sense to pop out of the `expr` state, I don't want to leave the other states that `include('expr')`.

I'll try separating the actual state from the included group.

JonathonReinhart replied:
That seemed to do the trick. See b3b1e66. (We can squash it if you're happy with it.)

This introduces a new state called `'assign-rhs'` which includes expressions, and is explicitly terminated on a newline. (I wish there was something like `default(blow up)` to assert this.)

JonathonReinhart replied:
Squashed.

jeanas commented on pygments/lexers/soong.py:
```diff
@@ -0,0 +1,77 @@
+            include('in-map'),
+        ],
+        'comments': [
+            (r'//.*?$', Comment.Single),
```
Unnecessary:

```suggestion
            (r'//.*?', Comment.Single),
```

JonathonReinhart replied:
Removing `$` seems to break everything.

I think the reason is because `.*?` is non-greedy. If you take away the subsequent character, it will match nothing:

https://docs.python.org/3/howto/regex.html#greedy-versus-non-greedy

```Python Console
>>> re.match(r'//.*?', '// foo!')
<re.Match object; span=(0, 2), match='//'>
>>> 
>>> re.match(r'//.*?!', '// foo!')
<re.Match object; span=(0, 7), match='// foo!'>
```

---

I took this from one of the other lexers. If you search, there are others exactly like this:
```console
$ git grep "\/\/\.\*\?\$', Comment"
```

JonathonReinhart replied:
Oh in fact, this comes directly from [the docs](https://pygments.org/docs/lexerdevelopment/#changing-states).

jeanas replied:
Oh, sorry, of course I meant to also remove the `?` at the end (and I've just made that change in the docs).

jeanas replied:
Change applied.

JonathonReinhart replied:
Got it. I made the change with no impact to the output.

jeanas commented on pygments/lexers/soong.py:
```diff
@@ -0,0 +1,77 @@
+            include('expr'),
+            include('comments'),
+            (r'\s+', Whitespace),  # newlines okay in a map
+            (r',', Punctuation),
```
You already have this in `'expr'`.

JonathonReinhart replied:
Actually, I think it probably makes sense to keep this one and remove the `,` from `expr`. (I think each item in a comma-separated list is an expression; the commas are between each expression).

I removed both `,` and 'comments' from `expr` with no breakage.

jeanas commented on pygments/lexers/soong.py:
```diff
@@ -0,0 +1,77 @@
+            include('expr'),
+            include('comments'),
+            (r'\s+', Whitespace),  # newlines okay in a list
+            (r',', Punctuation),
```
Already in `'expr'`

JonathonReinhart replied:
See [other comment](https://github.com/pygments/pygments/pull/2659/files/67edcd927cb0d61d5d54d874e9074ce085799508#r1544492882).

JonathonReinhart commented:
Thanks @jeanas. All comments have been resolved.

jeanas reviewed (approved):
Looks good. Thanks!
